File: src/dere_ambient/triggers/entities.py
"""Detect unfamiliar entities from newly created knowledge graph nodes."""

from __future__ import annotations

import re
from typing import Any

from .types import CuriositySignal

_GENERIC_ENTITY_NAMES = {
    "user",
    "assistant",
    "ai",
    "system",
    "daemon",
}


def detect_unfamiliar_entities(
    *,
    prompt: str,
    nodes: list[Any] | None,
    speaker_name: str | None,
    personality: str | None,
    max_entities: int = 3,
) -> list[CuriositySignal]:
    if not nodes or not prompt.strip():
        return []

    signals: list[CuriositySignal] = []
    for node in nodes:
        name = str(getattr(node, "name", "") or "").strip()
        if not name:
            continue

        if _is_generic_entity(node, name, speaker_name, personality):
            continue
        if _appears_as_log_prefix(name, prompt):
            continue

        signals.append(
            CuriositySignal(
                curiosity_type="unfamiliar_entity",
                topic=name,
                source_context=_truncate(prompt, 400),
                trigger_reason="New entity extracted from user message",
                user_interest=0.4,
            )
        )

        if len(signals) >= max_entities:
            break

    return signals


def _is_generic_entity(
    node: Any,
    name: str,
    speaker_name: str | None,
    personality: str | None,
) -> bool:
    if len(name) < 3:
        return True

    normalized = name.casefold()
    if normalized in _GENERIC_ENTITY_NAMES:
        return True

    labels = {str(label).lower() for label in getattr(node, "labels", [])}
    if labels & {"user", "assistant", "ai"}:
        return True

    if speaker_name and normalized == speaker_name.strip().lower():
        return True

    if personality and normalized == personality.strip().lower():
        return True

    return False


def _appears_as_log_prefix(name: str, prompt: str) -> bool:
    normalized = name.casefold()
    if not normalized:
        return False

    pattern = re.compile(rf"^\s*{re.escape(normalized)}(?:\.\d+)?\s*\|")
    for line in prompt.splitlines():
        if pattern.search(line.casefold()):
            return True
    return False


def _truncate(text: str, limit: int) -> str:
    if len(text) <= limit:
        return text
    return f"{text[:limit]}..."

File: packages/daemon/src/ambient-triggers/entities.ts
const GENERIC_ENTITY_NAMES = new Set(["user", "assistant", "ai", "system", "daemon"]);

export interface EntityNodeLike {
  name?: string | null;
  labels?: string[] | null;
}

export interface EntitySignal {
  topic: string;
  sourceContext: string;
  triggerReason: string;
}

export function detectUnfamiliarEntities(options: {
  prompt: string;
  nodes: EntityNodeLike[] | null;
  speakerName: string | null;
  personality: string | null;
  maxEntities?: number;
}): EntitySignal[] {
  const { prompt, nodes, speakerName, personality } = options;
  const maxEntities = options.maxEntities ?? 3;

  if (!nodes || prompt.trim().length === 0) {
    return [];
  }

  const signals: EntitySignal[] = [];
  for (const node of nodes) {
    const name = String(node.name ?? "").trim();
    if (!name) {
      continue;
    }

    if (isGenericEntity(node, name, speakerName, personality)) {
      continue;
    }
    if (appearsAsLogPrefix(name, prompt)) {
      continue;
    }

    signals.push({
      topic: name,
      sourceContext: truncate(prompt, 400),
      triggerReason: "New entity extracted from user message",
    });

    if (signals.length >= maxEntities) {
      break;
    }
  }

  return signals;
}

function isGenericEntity(
  node: EntityNodeLike,
  name: string,
  speakerName: string | null,
  personality: string | null,
): boolean {
  if (name.length < 3) {
    return true;
  }

  const normalized = name.toLowerCase();
  if (GENERIC_ENTITY_NAMES.has(normalized)) {
    return true;
  }

  const labels = new Set((node.labels ?? []).map((label) => String(label).toLowerCase()));
  if (["user", "assistant", "ai"].some((label) => labels.has(label))) {
    return true;
  }

  if (speakerName && normalized === speakerName.trim().toLowerCase()) {
    return true;
  }

  if (personality && normalized === personality.trim().toLowerCase()) {
    return true;
  }

  return false;
}

function appearsAsLogPrefix(name: string, prompt: string): boolean {
  const normalized = name.toLowerCase();
  if (!normalized) {
    return false;
  }
  const pattern = new RegExp(`^\\s*${escapeRegExp(normalized)}(?:\\.\\d+)?\\s*\\|`, "i");
  return prompt.split("\n").some((line) => pattern.test(line));
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function truncate(text: string, limit: number): string {
  if (text.length <= limit) {
    return text;
  }
  return `${text.slice(0, limit)}...`;
}
